File: R/pool.R
# call_metrics compiles a governed query over dictionary metrics, in the
# shape semantic layers converge on: metrics x dimensions x filters, plus
# simple column predicates. Names are passed as strings and validated at
# call time.
call_metrics_impl <- function(
  registry,
  sources,
  handles,
  metrics,
  dimensions = NULL,
  filters = NULL,
  where = NULL,
  source_name = NULL,
  semantic_models = NULL,
  arguments = "{}"
) {
  source <- resolve_sql_source(sources, source_name)
  label <- source_name %||% rlang::names2(sources)[[1]]
  n_models <- length(source$semantic_models)
  source <- source_hydrate_semantic_models(source, metrics)
  if (length(source$semantic_models) > n_models) {
    source_index <- if (length(sources) == 1L) {
      1L
    } else {
      match(label, names(sources))
    }
    sources[[source_index]] <- source
    semantic_models <- semantic_models_registry(sources)
  }
  defs <- registry_defs(registry, label)
  semantic_models <- semantic_models %||% semantic_models_registry(sources)
  semantic_members <- registry_semantic_members(semantic_models, label)
  origins <- resolve_metric_origins(metrics, defs, semantic_members)
  if (length(unique(origins)) > 1L) {
    cli::cli_abort(
      "Metrics in one call must come from either data-dict definitions or one native semantic model."
    )
  }
  if (identical(origins[[1]], "semantic_model")) {
    return(call_semantic_metrics(
      source,
      handles,
      semantic_members,
      metrics,
      dimensions,
      filters,
      where,
      source_name,
      arguments
    ))
  }

  if (length(parse_json_args(arguments))) {
    cli::cli_abort(
      "{.arg arguments} is supported only for native semantic metrics."
    )
  }

  metric_defs <- resolve_pool_names(metrics, defs, kind = "metric")
  if (any(metric_defs$mixed_grain)) {
    cli::cli_abort(
      "Mixed-grain metric{?s} {.val {metric_defs$name[metric_defs$mixed_grain]}} cannot be queried because the dependency chain requires a subquery rewrite."
    )
  }
  tables <- unique(metric_defs$table)
  if (length(tables) > 1) {
    cli::cli_abort(
      c(
        "Metrics in one query must share a table; these span {.val {tables}}.",
        i = if (is.null(handles)) {
          "Use {.fn run_sql} when documented relationships support the calculation."
        } else {
          "Query them separately and combine the results with run_r."
        }
      )
    )
  }
  on_table <- defs[defs$table == tables, ]
  columns <- names(source$dictionary$tables[[tables]]$columns)
  con <- source$con
  id <- DBI::dbQuoteIdentifier(con, source$table_ids[[tables]])

  dim_names <- strip_token_braces(dimensions %||% character())
  dims <- vapply(
    rlang::set_names(dim_names),
    function(name) dimension_sql(name, on_table, columns, con),
    character(1)
  )
  filter_defs <- resolve_pool_names(filters, on_table, kind = "filter")
  if (any(filter_defs$mixed_grain)) {
    cli::cli_abort(
      "Mixed-grain filter{?s} {.val {filter_defs$name[filter_defs$mixed_grain]}} cannot be applied by {.fn call_metrics}; use the definition in {.fn run_sql}."
    )
  }
  conditions <- c(
    sprintf("(%s)", filter_defs$sql),
    vapply(
      normalize_where(where),
      function(triple) compile_where_triple(triple, columns, con),
      character(1)
    )
  )

  select <- c(
    sprintf("%s AS %s", dims, DBI::dbQuoteIdentifier(con, names(dims))),
    sprintf(
      "(%s) AS %s",
      metric_defs$sql,
      DBI::dbQuoteIdentifier(con, metric_defs$name)
    )
  )
  sql <- sprintf(
    "SELECT %s FROM %s",
    paste(select, collapse = ", "),
    id
  )
  if (length(conditions)) {
    sql <- sprintf("%s WHERE %s", sql, paste(conditions, collapse = " AND "))
  }
  if (length(dims)) {
    sql <- sprintf("%s GROUP BY %s", sql, paste(dims, collapse = ", "))
  } else {
    # Force a global group so constants remain scalar when no rows remain.
    sql <- sprintf("%s HAVING COUNT(*) >= 0", sql)
  }

  result <- source_query(source, sql)
  advert <- register_handle(handles, result)
  dimension_defs <- on_table[
    match(dim_names, on_table$name, nomatch = 0L),
  ]
  applied <- rbind(metric_defs, dimension_defs, filter_defs)
  applied <- applied[!duplicated(applied[c("name", "table", "source")]), ]
  note <- applied_definitions_text(applied)
  args <- drop_nulls(list(
    metrics = metrics,
    dimensions = dimensions,
    filters = filters
  ))
  metric_tool_result(
    result,
    sql,
    handles,
    metrics,
    dimensions,
    filters,
    source_name,
    note = note,
    advert = advert,
    args = args
  )
}

resolve_metric_origins <- function(metrics, definitions, semantic_members) {
  metrics <- strip_token_braces(metrics %||% character())
  if (length(metrics) == 0L) {
    cli::cli_abort("{.arg metrics} must contain at least one metric name.")
  }
  vapply(
    metrics,
    metric_origin,
    character(1),
    definitions = definitions,
    semantic_members = semantic_members
  )
}

metric_origin <- function(name, definitions, semantic_members) {
  definition_named <- pool_name_candidates(name, definitions)
  definition_metrics <- definition_named[
    definition_named$kind == "metric",
    ,
    drop = FALSE
  ]
  semantic_named <- semantic_member_candidates(name, semantic_members)
  semantic_metrics <- semantic_named[
    semantic_named$kind == "metric",
    ,
    drop = FALSE
  ]
  n_matches <- nrow(definition_metrics) + nrow(semantic_metrics)
  if (n_matches == 1L) {
    if (nrow(definition_metrics)) "definition" else "semantic_model"
  } else if (n_matches > 1L) {
    definition_choices <- sprintf(
      "%s::%s",
      definition_metrics$table,
      definition_metrics$name
    )
    semantic_choices <- vapply(seq_len(nrow(semantic_metrics)), function(i) {
      semantic_member_key(semantic_metrics[i, , drop = FALSE])
    }, character(1))
    cli::cli_abort(c(
      "Metric name {.val {name}} is ambiguous.",
      "i" = "Use a qualified name: {.val {c(definition_choices, semantic_choices)}}."
    ))
  } else if (nrow(definition_named) || nrow(semantic_named)) {
    kinds <- unique(c(definition_named$kind, semantic_named$kind))
    kind_text <- paste(kinds, collapse = " or a ")
    cli::cli_abort(
      "{.val {name}} is a {kind_text}, not a metric."
    )
  } else {
    available <- c(
      definitions$name[definitions$kind == "metric"],
      semantic_members$name[semantic_members$kind == "metric"]
    )
    cli::cli_abort(c(
      "No governed metric is named {.val {name}}.",
      "i" = "Available metrics: {.val {available}}."
    ))
  }
}

call_semantic_metrics <- function(
  source,
  handles,
  members,
  metrics,
  dimensions,
  filters,
  where,
  source_name,
  arguments
) {
  metric_members <- resolve_semantic_members(metrics, members, "metric")
  models <- unique(metric_members$model)
  if (length(models) != 1L) {
    cli::cli_abort("Metrics in one call must belong to one native semantic model.")
  }
  model <- source$semantic_models[[models[[1]]]]
  model_members <- members[members$model == models[[1]], , drop = FALSE]
  dimension_members <- resolve_semantic_members(
    dimensions,
    model_members,
    "dimension"
  )
  filter_members <- resolve_semantic_filters(filters, model_members)
  arguments <- validate_typed_arguments(
    model$parameters,
    parse_json_args(arguments),
    include_defaults = FALSE
  )
  sql <- switch(
    model$backend,
    snowflake_semantic_view = snowflake_semantic_metric_sql(
      model,
      metric_members,
      dimension_members,
      filter_members,
      where,
      model_members,
      source$con,
      arguments = arguments
    ),
    databricks_metric_view = {
      if (nrow(filter_members)) {
        cli::cli_abort(
          "Named filters from Databricks metric views are not supported."
        )
      }
      databricks_semantic_metric_sql(
        model,
        metric_members,
        dimension_members,
        where,
        model_members,
        source$con,
        arguments = arguments
      )
    },
    cli::cli_abort(
      "Unsupported native semantic-model backend {.val {model$backend}}."
    )
  )
  result <- source_query_bind(source, sql, unname(arguments))
  metric_tool_result(
    result,
    sql,
    handles,
    metrics,
    dimensions,
    filters,
    source_name
  )
}

metric_tool_result <- function(
  result,
  sql,
  handles,
  metrics,
  dimensions,
  filters,
  source_name,
  note = NULL,
  advert = register_handle(handles, result),
  args = drop_nulls(list(
    metrics = metrics,
    dimensions = dimensions,
    filters = filters
  ))
) {
  tool_result(
    paste(c(df_to_markdown(result), note, advert), collapse = "\n\n"),
    title = sprintf(
      "Metrics: %s%s",
      html_escape(paste(metrics, collapse = ", ")),
      source_label(source_name)
    ),
    icon = maybe_icon("shield-check"),
    markdown = sprintf("```sql\n%s\n```\n\n%s", sql, df_to_markdown(result)),
    html = measure_display_html(args, result),
    tag = "A",
    show_tag = FALSE
  )
}

registry_has_metrics <- function(registry) {
  any(registry$defs$kind == "metric")
}

# The prompt teaches `{{name}}` for SQL, so models sometimes pass the
# braces here too; accept both forms.
strip_token_braces <- function(name) {
  gsub("^\\{\\{\\s*|\\s*\\}\\}$", "", trimws(name))
}

resolve_pool_names <- function(names, defs, kind) {
  out <- defs[0, ]
  for (name in strip_token_braces(names %||% character())) {
    out <- rbind(out, resolve_pool_name(name, defs, kind))
  }
  out
}

resolve_pool_name <- function(name, defs, kind) {
  named <- pool_name_candidates(name, defs)
  matched <- named[which(named$kind == kind), ]
  if (nrow(matched) == 1L) {
    return(matched[1, ])
  }
  if (nrow(matched) > 1L) {
    qualified <- sprintf("{{%s::%s}}", matched$table, matched$name)
    cli::cli_abort(
      c(
        "Governed {kind} name {.val {name}} is ambiguous.",
        i = "Qualify it as {.or {.code {qualified}}}."
      )
    )
  }
  if (nrow(named)) {
    cli::cli_abort(
      "{.val {name}} is a {named$kind[[1]]}, not a {kind}; apply it as
       {.code {{{{{name}}}}}} in SQL instead."
    )
  }
  available <- defs$name[which(defs$kind == kind)]
  cli::cli_abort(
    c(
      "No governed {kind} is named {.val {name}}.",
      i = "Available {kind}s: {.val {available}}."
    )
  )
}

pool_name_candidates <- function(name, defs) {
  separator <- regexpr("::", name, fixed = TRUE)[[1]]
  if (separator > 0L) {
    table <- substr(name, 1L, separator - 1L)
    definition <- substr(name, separator + 2L, nchar(name))
    return(defs[defs$table == table & defs$name == definition, ])
  }
  defs[defs$name == name, ]
}

dimension_sql <- function(name, defs, columns, con) {
  named <- defs[defs$name == name, ]
  if (nrow(named)) {
    if (!named$kind[[1]] %in% c("derived", "filter")) {
      cli::cli_abort(
        "{.val {name}} is a {named$kind[[1]]} and can't be grouped by."
      )
    }
    if (named$mixed_grain[[1]]) {
      cli::cli_abort(
        "Mixed-grain definition {.val {name}} cannot be grouped by with {.fn call_metrics}; use it in {.fn run_sql}."
      )
    }
    return(sprintf("(%s)", named$sql[[1]]))
  }
  if (name %in% columns) {
    return(as.character(DBI::dbQuoteIdentifier(con, name)))
  }
  dimensions <- defs$name[
    which(defs$kind %in% c("derived", "filter") & !defs$mixed_grain)
  ]
  cli::cli_abort(
    c(
      "No dimension or documented column is named {.val {name}}.",
      i = "Documented columns: {.val {columns}}.",
      i = if (length(dimensions)) {
        "Governed row definitions: {.val {dimensions}}."
      }
    )
  )
}

# ellmer may deliver an array of objects as a data frame or a list of named
# lists depending on the provider payload; normalize to a list of triples.
normalize_where <- function(where) {
  if (is.null(where) || length(where) == 0) {
    return(list())
  }
  if (is.data.frame(where)) {
    return(lapply(seq_len(nrow(where)), function(i) as.list(where[i, ])))
  }
  if (!is.null(names(where))) {
    return(list(as.list(where)))
  }
  lapply(where, as.list)
}

where_ops <- c("=", "!=", "<", "<=", ">", ">=")

compile_where_triple <- function(triple, columns, con) {
  for (field in c("column", "op", "value")) {
    value <- triple[[field]]
    if (length(value) != 1 || is.na(value) || !nzchar(as.character(value))) {
      cli::cli_abort(
        "Each {.arg where} entry needs {.field column}, {.field op}, and
         {.field value}."
      )
    }
    triple[[field]] <- as.character(value)
  }
  if (!triple$column %in% columns) {
    cli::cli_abort(
      c(
        "{.arg where} references {.val {triple$column}}, which is not a
         documented column.",
        i = "Documented columns: {.val {columns}}."
      )
    )
  }
  if (!triple$op %in% where_ops) {
    cli::cli_abort(
      "{.arg where} operator must be one of {.val {where_ops}},
       not {.val {triple$op}}."
    )
  }
  value <- if (grepl("^-?[0-9]+(\\.[0-9]+)?$", triple$value)) {
    triple$value
  } else {
    as.character(DBI::dbQuoteString(con, triple$value))
  }
  sprintf(
    "(%s %s %s)",
    DBI::dbQuoteIdentifier(con, triple$column),
    triple$op,
    value
  )
}

# One discovery surface over the whole pool: measures and definitions ranked
# together, so the model doesn't have to guess which kind holds its answer.
search_pool_text <- function(
  measures,
  registry,
  query,
  source_names = character(),
  semantic_models = NULL,
  calculations = list()
) {
  defs <- registry_defs(registry)
  semantic_models <- semantic_models %||% list(members = no_semantic_members)
  semantic_members <- registry_semantic_members(semantic_models)
  semantic_stubs <- registry_semantic_stubs(semantic_models)
  if (
    length(measures) == 0 &&
      nrow(defs) == 0 &&
      nrow(semantic_members) == 0 &&
      nrow(semantic_stubs) == 0 &&
      length(calculations) == 0
  ) {
    return("The semantic layer is empty.")
  }

  blank_na <- function(x) ifelse(is.na(x), "", x)
  catalog <- c(
    vapply(
      measures,
      function(td) paste(tool_name(td), tool_description(td)),
      character(1)
    ),
    paste(
      defs$name,
      defs$table,
      defs$kind,
      blank_na(defs$label),
      blank_na(defs$description),
      blank_na(defs$details),
      blank_na(defs$sql)
    ),
    paste(
      semantic_members$name,
      semantic_members$model,
      semantic_members$kind,
      blank_na(semantic_members$label),
      blank_na(semantic_members$description),
      semantic_members$synonyms
    ),
    paste(
      semantic_stubs$name,
      semantic_stubs$model,
      semantic_stubs$backend,
      blank_na(semantic_stubs$description)
    ),
    vapply(
      calculations,
      function(calculation) paste(
        calculation$key,
        calculation$name,
        calculation$description,
        paste(
          vapply(
            calculation$arguments,
            function(argument) argument$description %||% "",
            character(1)
          ),
          collapse = " "
        )
      ),
      character(1)
    )
  )

  hits <- lexical_rank(query, catalog, n = 5)
  if (length(hits) == 0) {
    return(sprintf(
      "Nothing in the semantic layer matches \"%s\". Consider writing a SQL query.",
      query
    ))
  }
  blocks <- vapply(
    hits,
    function(hit) {
      if (hit <= length(measures)) {
        measure_schema_text(measures[[hit]], source_names = source_names)
      } else if (hit <= length(measures) + nrow(defs)) {
        definition_pool_text(defs[hit - length(measures), ], defs)
      } else if (
        hit <= length(measures) + nrow(defs) + nrow(semantic_members)
      ) {
        semantic_member_pool_text(
          semantic_members[hit - length(measures) - nrow(defs), , drop = FALSE],
          semantic_members,
          source_names,
          registry_semantic_parameters(
            semantic_models,
            semantic_members[
              hit - length(measures) - nrow(defs),
              ,
              drop = FALSE
            ]
          )
        )
      } else if (
        hit <= length(measures) +
          nrow(defs) +
          nrow(semantic_members) +
          nrow(semantic_stubs)
      ) {
        semantic_stub_pool_text(
          semantic_stubs[
            hit - length(measures) - nrow(defs) - nrow(semantic_members),
            ,
            drop = FALSE
          ],
          source_names
        )
      } else {
        calculation_pool_text(
          calculations[[
            hit -
              length(measures) -
              nrow(defs) -
              nrow(semantic_members) -
              nrow(semantic_stubs)
          ]],
          source_names
        )
      }
    },
    character(1)
  )
  paste(blocks, collapse = "\n\n")
}

semantic_stub_pool_text <- function(stub, source_names = character()) {
  backend <- switch(
    stub$backend[[1]],
    snowflake_semantic_view = "Snowflake semantic view",
    databricks_metric_view = "Databricks metric view",
    "semantic model"
  )
  source <- if (stub$source[[1]] %in% source_names) {
    sprintf("\nSource: `%s`.", stub$source[[1]])
  } else {
    ""
  }
  sprintf(
    paste0(
      "### %s --- %s\n%s%s\n",
      if (identical(stub$backend[[1]], "snowflake_semantic_view")) {
        paste(
          "Inspect with `describe_table` to find its public members and",
          "verified queries."
        )
      } else {
        "Inspect with `describe_table` to find its public members."
      }
    ),
    stub$model[[1]],
    backend,
    prose_detail(stub$description[[1]], NA_character_),
    source
  )
}

definition_pool_text <- function(def, defs) {
  kind <- def$kind
  table_has_metrics <- any(
    defs$source == def$source &
      defs$table == def$table &
      defs$kind == "metric"
  )
  invoke <- if (def$mixed_grain) {
    sprintf(
      "Use `{{%s}}` only in run_sql with manually grain-correct query structure.",
      def$name
    )
  } else {
    switch(
      kind,
      filter = if (table_has_metrics) {
        sprintf(
          "Apply in run_sql (e.g. `WHERE {{%s}}`) or as a call_metrics filter or dimension.",
          def$name
        )
      } else {
        sprintf("Apply in run_sql (e.g. `WHERE {{%s}}`).", def$name)
      },
      metric = sprintf(
        "Query with call_metrics (metrics = [\"%s\"]) or in run_sql as `SELECT {{%s}} AS value`.",
        def$name,
        def$name
      ),
      if (table_has_metrics) {
        sprintf(
          "Use in run_sql SELECT or GROUP BY as `{{%s}}`, or as a call_metrics dimension.",
          def$name
        )
      } else {
        sprintf(
          "Use in run_sql SELECT or GROUP BY as `{{%s}}`.",
          def$name
        )
      }
    )
  }

  siblings <- NULL
  if (identical(kind, "metric")) {
    same_table <- defs[
      defs$table == def$table &
        defs$name != def$name &
        defs$kind %in% c("filter", "derived") &
        !defs$mixed_grain,
    ]
    if (nrow(same_table)) {
      items <- sprintf(
        "{{%s}} (%s)",
        same_table$name,
        same_table$kind
      )
      siblings <- sprintf(
        "Filters and derived definitions on this table: %s.",
        paste(items, collapse = ", ")
      )
    }
  }

  paste(
    c(
      sprintf(
        "### {{%s}} --- %s on table `%s`\n%s",
        def$name,
        kind,
        def$table,
        prose_detail(def$description, def$details)
      ),
      sprintf(
        "Selected %s: `(%s)`.",
        def$target,
        flatten_inline(def$sql)
      ),
      if (length(def$notes[[1]])) {
        sprintf(
          "Translation notes: %s",
          paste(def$notes[[1]], collapse = " ")
        )
      },
      invoke,
      siblings
    ),
    collapse = "\n"
  )
}

File: pkg-py/src/commons/_pool.py
"""The pool: one discovery surface over an agent's trusted calculations, and
the governed query `call_metrics` compiles.

Measures and governed definitions are ranked together, so the model does not
have to guess which kind holds its answer. `pkg-r/R/pool.R` decides the same
two things for R.
"""

from __future__ import annotations

import re
from collections.abc import Mapping, Sequence
from typing import TYPE_CHECKING, Any

from chatlas import ContentToolResult
from sqlglot import expressions as exp

from ._citations import tool_result
from ._definitions import ExportRecord, applied_text
from ._measures import Measure, measure_schema_text
from ._provenance import Tag
from ._rows import rows_to_markdown
from ._sql_guard import sqlglot_dialect

if TYPE_CHECKING:
    from ._data_source import DataSource
    from ._definitions import Registry
    from ._handles import HandleStore

__all__ = [
    "WHERE_OPERATORS",
    "call_metrics",
    "definition_pool_text",
    "lexical_rank",
    "search_pool_text",
]

WHERE_OPERATORS = ("=", "!=", "<", "<=", ">", ">=")

_NOT_WORD = re.compile(r"[^a-z0-9]+")
_TOKEN_BRACES = re.compile(r"^\{\{\s*|\s*\}\}$")
_NUMERIC = re.compile(r"^-?[0-9]+(\.[0-9]+)?$")


def _tokenize(text: str) -> list[str]:
    # Single characters are dropped: they match too much to rank anything.
    return [term for term in _NOT_WORD.split(text.lower()) if len(term) > 1]


def lexical_rank(query: str, documents: Sequence[str], limit: int = 5) -> list[int]:
    """The positions of the documents sharing the most terms with `query`.

    A term repeated in the query counts each time, and a term repeated in a
    document counts once, so a document is ranked by how much of the query it
    covers rather than by how often it says one word.
    """
    terms = _tokenize(query)
    if not terms or not documents:
        return []
    scored = [
        (sum(term in set(_tokenize(document)) for term in terms), position)
        for position, document in enumerate(documents)
    ]
    hits = [entry for entry in scored if entry[0] > 0]
    # A stable sort on the score alone, so equally scored documents stay in
    # the order the pool presents them.
    hits.sort(key=lambda entry: entry[0], reverse=True)
    return [position for _, position in hits[:limit]]


def _flatten_inline(text: str) -> str:
    return re.sub(r"\s*\n\s*", " ", text).strip()


def _prose_detail(description: str | None, details: str | None) -> str:
    return _flatten_inline(" ".join(part for part in (description, details) if part))


def _strip_token_braces(name: str) -> str:
    # The prompt teaches `{{name}}` for SQL, so models sometimes pass the
    # braces here too; accept both forms.
    return _TOKEN_BRACES.sub("", name.strip())


def search_pool_text(
    measures: Mapping[str, Measure],
    registry: Registry,
    query: str,
    source_names: Sequence[str] = (),
) -> str:
    """The blocks `search_pool` shows for the best matches in the pool."""
    records = registry.records
    entries = list(measures.values())
    if not entries and not records:
        return "The semantic layer is empty."

    catalog = [f"{record.name} {record.description}" for record in entries]
    catalog += [
        " ".join(
            part or ""
            for part in (
                record.name,
                record.table,
                record.kind,
                record.label,
                record.description,
                record.details,
                record.sql,
            )
        )
        for record in records
    ]

    hits = lexical_rank(query, catalog)
    if not hits:
        return (
            f'Nothing in the semantic layer matches "{query}". '
            "Consider writing a SQL query."
        )
    blocks = [
        measure_schema_text(entries[hit], source_names=source_names)
        if hit < len(entries)
        else definition_pool_text(records[hit - len(entries)], records)
        for hit in hits
    ]
    return "\n\n".join(blocks)


def definition_pool_text(record: ExportRecord, records: Sequence[ExportRecord]) -> str:
    """One governed definition as the pool presents it, with how to invoke it.

    Sibling definitions on the same table are named under a metric, because a
    metric is what a grouped or filtered query is built around.
    """
    name = record.name
    table_has_metrics = any(
        other.source == record.source
        and other.table == record.table
        and other.kind == "metric"
        for other in records
    )

    if record.mixed_grain:
        invoke = (
            f"Use `{{{{{name}}}}}` only in run_sql with manually grain-correct "
            "query structure."
        )
    elif record.kind == "filter":
        invoke = f"Apply in run_sql (e.g. `WHERE {{{{{name}}}}}`)"
        invoke += (
            " or as a call_metrics filter or dimension." if table_has_metrics else "."
        )
    elif record.kind == "metric":
        invoke = (
            f'Query with call_metrics (metrics = ["{name}"]) or in run_sql as '
            f"`SELECT {{{{{name}}}}} AS value`."
        )
    else:
        invoke = f"Use in run_sql SELECT or GROUP BY as `{{{{{name}}}}}`"
        invoke += ", or as a call_metrics dimension." if table_has_metrics else "."

    heading = f"### {{{{{name}}}}} --- {record.kind} on table `{record.table}`"
    parts = [
        f"{heading}\n{_prose_detail(record.description, record.details)}",
        f"Selected {record.target}: `({_flatten_inline(record.sql)})`.",
    ]
    if record.notes:
        parts.append(f"Translation notes: {' '.join(record.notes)}")
    parts.append(invoke)

    if record.kind == "metric":
        siblings = [
            other
            for other in records
            # Scoped to this source: a sibling on another source's table of
            # the same name is not reachable from a call_metrics query
            # against this one. pkg-r/R/pool.R lists siblings across sources;
            # the scoping here is deliberate, not shared behavior.
            if other.source == record.source
            and other.table == record.table
            and other.name != record.name
            and other.kind in ("filter", "derived")
            and not other.mixed_grain
        ]
        if siblings:
            listed = ", ".join(
                f"{{{{{other.name}}}}} ({other.kind})" for other in siblings
            )
            parts.append(f"Filters and derived definitions on this table: {listed}.")
    return "\n".join(parts)


# ---- the governed query ---------------------------------------------------


def call_metrics(
    registry: Registry,
    source: DataSource,
    label: str,
    handles: HandleStore | None,
    metrics: list[str],
    dimensions: list[str] | None = None,
    filters: list[str] | None = None,
    where: list[dict[str, Any]] | None = None,
) -> ContentToolResult:
    """Compile and run a governed query over one table's definitions.

    The shape semantic layers converge on: metrics by dimensions with
    filters, plus simple column predicates. Names arrive as strings and are
    validated here, because the provider only ever sees `call_metrics`.
    """
    if not metrics:
        raise ValueError("metrics must name at least one governed metric.")
    defs = registry.for_source(label)
    metric_records = _resolve_names(metrics, defs, "metric")
    _refuse_mixed_grain(
        metric_records,
        "metric",
        "cannot be queried because the dependency chain requires a subquery rewrite",
    )

    tables = list(dict.fromkeys(record.table for record in metric_records))
    if len(tables) > 1:
        raise ValueError(
            f"Metrics in one query must share a table; these span "
            f"{', '.join(tables)}. Query them separately and combine the "
            "results with run_python."
        )
    table = tables[0]
    on_table = [record for record in defs if record.table == table]

    dictionary = source.dictionary
    entry = dictionary.tables.get(table) if dictionary is not None else None
    columns = list(entry.columns) if entry is not None else []
    dialect = sqlglot_dialect(source.backend.dialect())

    dimension_names = [_strip_token_braces(name) for name in dimensions or []]
    dimension_sql = {
        name: _dimension_sql(name, on_table, columns, dialect)
        for name in dimension_names
    }
    filter_records = _resolve_names(filters, on_table, "filter")
    _refuse_mixed_grain(
        filter_records,
        "filter",
        "cannot be applied by call_metrics; use the definition in run_sql",
    )

    conditions = [f"({record.sql})" for record in filter_records]
    conditions += [_where_condition(triple, columns, dialect) for triple in where or []]

    select = [
        f"{sql} AS {_quote_name(name, dialect)}" for name, sql in dimension_sql.items()
    ]
    select += [
        f"({record.sql}) AS {_quote_name(record.name, dialect)}"
        for record in metric_records
    ]
    sql = f"SELECT {', '.join(select)} FROM {source.backend.quote(source.table_ids[table])}"
    if conditions:
        sql += f" WHERE {' AND '.join(conditions)}"
    if dimension_sql:
        sql += f" GROUP BY {', '.join(dimension_sql.values())}"
    else:
        # Force a global group, so a constant stays scalar when no rows remain.
        sql += " HAVING COUNT(*) >= 0"

    rows = source.query(sql)
    advert = handles.register(rows) if handles is not None else None
    applied: list[ExportRecord] = []
    for record in (
        *metric_records,
        *(record for record in on_table if record.name in dimension_names),
        *filter_records,
    ):
        if record not in applied:
            applied.append(record)
    body = "\n\n".join(
        part for part in (rows_to_markdown(rows), applied_text(applied), advert) if part
    )
    return tool_result(body, tag=Tag.A)


def _resolve_names(
    names: list[str] | None, defs: Sequence[ExportRecord], kind: str
) -> list[ExportRecord]:
    return [
        _resolve_name(_strip_token_braces(name), defs, kind) for name in names or []
    ]


def _resolve_name(name: str, defs: Sequence[ExportRecord], kind: str) -> ExportRecord:
    named = _candidates(name, defs)
    matched = [record for record in named if record.kind == kind]
    if len(matched) == 1:
        return matched[0]
    if matched:
        qualified = " or ".join(
            f"{{{{{record.table}::{record.name}}}}}" for record in matched
        )
        raise ValueError(
            f"Governed {kind} name {name!r} is ambiguous. Qualify it as {qualified}."
        )
    if named:
        raise ValueError(
            f"{name!r} is a {named[0].kind}, not a {kind}; apply it as "
            f"{{{{{name}}}}} in SQL instead."
        )
    available = [record.name for record in defs if record.kind == kind]
    raise ValueError(
        f"No governed {kind} is named {name!r}. "
        f"Available {kind}s: {', '.join(available)}."
    )


def _candidates(name: str, defs: Sequence[ExportRecord]) -> list[ExportRecord]:
    if "::" in name:
        table, _, definition = name.partition("::")
        return [
            record
            for record in defs
            if record.table == table and record.name == definition
        ]
    return [record for record in defs if record.name == name]


def _refuse_mixed_grain(
    records: Sequence[ExportRecord], kind: str, reason: str
) -> None:
    mixed = [record.name for record in records if record.mixed_grain]
    if mixed:
        noun = kind if len(mixed) == 1 else f"{kind}s"
        raise ValueError(f"Mixed-grain {noun} {', '.join(mixed)} {reason}.")


def _dimension_sql(
    name: str, defs: Sequence[ExportRecord], columns: list[str], dialect: str | None
) -> str:
    named = _candidates(name, defs)
    if named:
        record = named[0]
        if record.kind not in ("derived", "filter"):
            raise ValueError(f"{name!r} is a {record.kind} and can't be grouped by.")
        if record.mixed_grain:
            raise ValueError(
                f"Mixed-grain definition {name!r} cannot be grouped by with "
                "call_metrics; use it in run_sql."
            )
        return f"({record.sql})"
    if name in columns:
        return _quote_name(name, dialect)
    groupable = [
        record.name
        for record in defs
        if record.kind in ("derived", "filter") and not record.mixed_grain
    ]
    detail = f" Governed row definitions: {', '.join(groupable)}." if groupable else ""
    raise ValueError(
        f"No dimension or documented column is named {name!r}. "
        f"Documented columns: {', '.join(columns)}.{detail}"
    )


def _where_condition(triple: Any, columns: list[str], dialect: str | None) -> str:
    if not isinstance(triple, Mapping):
        raise TypeError("Each where entry needs column, op, and value, as an object.")
    values = {}
    for field in ("column", "op", "value"):
        value = triple.get(field)
        if value is None or not str(value):
            raise ValueError("Each where entry needs column, op, and value.")
        values[field] = str(value)
    if values["column"] not in columns:
        raise ValueError(
            f"where references {values['column']!r}, which is not a documented "
            f"column. Documented columns: {', '.join(columns)}."
        )
    if values["op"] not in WHERE_OPERATORS:
        raise ValueError(
            f"where operator must be one of {', '.join(WHERE_OPERATORS)}, not "
            f"{values['op']!r}."
        )
    # A number is written bare so the comparison stays numeric; anything else
    # is a literal the dialect quotes for itself.
    literal = (
        values["value"]
        if _NUMERIC.match(values["value"])
        else _quote_string(values["value"], dialect)
    )
    return f"({_quote_name(values['column'], dialect)} {values['op']} {literal})"


def _quote_name(name: str, dialect: str | None) -> str:
    return exp.to_identifier(name, quoted=True).sql(dialect=dialect)


def _quote_string(value: str, dialect: str | None) -> str:
    return exp.Literal.string(value).sql(dialect=dialect)
